8264591: HBox/VBox child widths pixel-snap to wrong value

mstr2 opened the pull request:
The children of HBox/VBox don't always pixel-snap to the same value as the container itself when a render scale other than 1 is used. This can lead to a visual glitch where the content bounds don't line up with the container bounds. In this case, the container will extend beyond its content, or vice versa.

The following program shows the problem for HBox:
```
Region r1 = new Region();
Region r2 = new Region();
Region r3 = new Region();
Region r4 = new Region();
Region r5 = new Region();
Region r6 = new Region();
r1.setBackground(new Background(new BackgroundFill(Color.GREY, null, null)));
r2.setBackground(new Background(new BackgroundFill(Color.DARKGRAY, null, null)));
r3.setBackground(new Background(new BackgroundFill(Color.BLACK, null, null)));
r4.setBackground(new Background(new BackgroundFill(Color.GREY, null, null)));
r5.setBackground(new Background(new BackgroundFill(Color.DARKGRAY, null, null)));
r6.setBackground(new Background(new BackgroundFill(Color.BLACK, null, null)));
r1.setPrefWidth(25.3);
r2.setPrefWidth(25.3);
r3.setPrefWidth(25.4);
r4.setPrefWidth(25.3);
r5.setPrefWidth(25.3);
r6.setPrefWidth(25.4);
r1.setMaxHeight(30);
r2.setMaxHeight(30);
r3.setMaxHeight(30);
r4.setMaxHeight(30);
r5.setMaxHeight(30);
r6.setMaxHeight(30);
HBox hbox1 = new HBox(r1, r2, r3, r4, r5, r6);
hbox1.setBackground(new Background(new BackgroundFill(Color.RED, null, null)));
hbox1.setPrefHeight(40);

r1 = new Region();
r2 = new Region();
r3 = new Region();
r4 = new Region();
r5 = new Region();
r6 = new Region();
r1.setBackground(new Background(new BackgroundFill(Color.GREY, null, null)));
r2.setBackground(new Background(new BackgroundFill(Color.DARKGRAY, null, null)));
r3.setBackground(new Background(new BackgroundFill(Color.BLACK, null, null)));
r4.setBackground(new Background(new BackgroundFill(Color.GREY, null, null)));
r5.setBackground(new Background(new BackgroundFill(Color.DARKGRAY, null, null)));
r6.setBackground(new Background(new BackgroundFill(Color.BLACK, null, null)));
r1.setPrefWidth(25.3);
r2.setPrefWidth(25.3);
r3.setPrefWidth(25.4);
r4.setPrefWidth(25.3);
r5.setPrefWidth(25.3);
r6.setPrefWidth(25.4);
r1.setMaxHeight(30);
r2.setMaxHeight(30);
r3.setMaxHeight(30);
r4.setMaxHeight(30);
r5.setMaxHeight(30);
r6.setMaxHeight(30);
HBox hbox2 = new HBox(r1, r2, r3, r4, r5, r6);
hbox2.setBackground(new Background(new BackgroundFill(Color.RED, null, null)));
hbox2.setPrefHeight(40);
hbox2.setPrefWidth(152);

VBox root = new VBox(new HBox(hbox1), new HBox(hbox2));
root.setSpacing(20);
Scene scene = new Scene(root, 500, 250);

primaryStage.setScene(scene);
primaryStage.show();
```

Here's a before-and-after comparison of the program above after applying the fix. Note that 'before', the backgrounds of the container (red) and its content (black) don't align perfectly. The render scale is 175% in this example.
![pixel-glitch](https://user-images.githubusercontent.com/43553916/112891996-146e2d80-90d9-11eb-9d83-97754ae38dc1.png)

I've filed a bug report and will change the title of the GitHub issue as soon as it's posted.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Integration blocker
&nbsp;⚠️ Too few reviewers with at least role reviewer found (have 0, need at least 1) (failed with the updated jcheck configuration)

### Issue
 * [JDK-8264591](https://bugs.openjdk.org/browse/JDK-8264591): HBox/VBox child widths pixel-snap to wrong value (**Bug** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/445/head:pull/445` \
`$ git checkout pull/445`

Update a local copy of the PR: \
`$ git checkout pull/445` \
`$ git pull https://git.openjdk.org/jfx.git pull/445/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 445`

View PR using the GUI difftool: \
`$ git pr show -t 445`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/445.diff">https://git.openjdk.org/jfx/pull/445.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/445#issuecomment-811922562)

kevinrushforth commented:
/reviewers 2

kevinrushforth commented:
The bug is posted here:

[JDK-8264591](https://bugs.openjdk.java.net/browse/JDK-8264591): HBox/VBox child widths pixel-snap to wrong value

Maran23 commented:
I had a look at the PR. But it took quite a while to fully understand the changes (also the current implementation 😄). 
I think it make sense to improve it a bit e.g. by adding javadoc to the new methods, maybe also the existing? Other ideas are also welcome. 
With that said maybe more people will review it then. I will have a full look soon as well. :)

mstr2 commented:
> I had a look at the PR. But it took quite a while to fully understand the changes (also the current implementation 😄). I think it make sense to improve it a bit e.g. by adding javadoc to the new methods, maybe also the existing? Other ideas are also welcome. With that said maybe more people will review it then. I will have a full look soon as well. :)

I've added a bit of documentation.

mstr2 commented:
@Maran23 Would you be interested in reviewing the implementation and the added documentation?

mstr2 commented:
I've expanded the documentation of the improved algorithm, which hopefully makes it easier to understand the changes.

Maran23 reviewed (commented):
Tested the changes with some easy and more complex layout and it looks good

mstr2 commented on modules/javafx.graphics/src/test/java/test/javafx/scene/layout/VBoxTest.java:
```diff
@@ -783,4 +786,75 @@ public class VBoxTest {
+        for (testPixelSnapConfig config : new testPixelSnapConfig[] {
+            // For these tests, VBox.prefHeight is specified, so we expect the final height to be exactly that.
+            // Child heights will be adjusted appropriately such that the sum of child widths corresponds to VBox.prefHeight.
+            new testPixelSnapConfig(76.0, 1.0, true),
```
Done.

mstr2 commented on modules/javafx.graphics/src/test/java/test/javafx/scene/layout/VBoxTest.java:
```diff
@@ -783,4 +786,75 @@ public class VBoxTest {
+     * ╚═══════════════════╝
+     */
+    @Test public void testPixelSnappedContentHeightIsSameAsBoxHeight() {
+        class testPixelSnapConfig {
```
I've renamed the class. Since it is not reusable and only relevant within this single test, I think it's okay to leave it as a local class.

mstr2 commented:
> @mstr2 Please do not rebase or force-push to an active PR as it invalidates existing review comments. Note for future reference, the bots always squash all changes into a single commit automatically as part of the integration. See [OpenJDK Developers’ Guide](https://openjdk.org/guide/#working-with-pull-requests) for more information.

Not sure what happened here. I didn't rebase or force-push.

mstr2 commented:
@Maran23 You previously reviewed and tested the proposed changes. If everything looks good to you, can you approve this PR?

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java:
```diff
@@ -467,77 +472,170 @@ private double[][] getAreaWidths(List<Node>managed, double height, boolean minim
+        // We repeatedly apply the following algorithm as long as we have space left to
+        // distribute (currentDelta), as well as children that are eligible to grow or
+        // shrink (adjustingNumber).
+        while ((currentDelta > Double.MIN_VALUE || currentDelta < -Double.MIN_VALUE) && adjustingNumber > 0) {
```
is there a guarantee that this while loop ends?

mstr2 replied:
Yes, in line 612 we return from this method once we can't resize any further.

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/javafx/scene/layout/VBox.java:
```diff
@@ -457,77 +460,166 @@ private double[][] getAreaHeights(List<Node>managed, double width, boolean minim
+        // We repeatedly apply the following algorithm as long as we have space left to
+        // distribute (currentDelta), as well as children that are eligible to grow or
+        // shrink (adjustingNumber).
+        while ((currentDelta > Double.MIN_VALUE || currentDelta < -Double.MIN_VALUE) && adjustingNumber > 0) {
```
same question: is it possible to enter an infinite loop?

I suspect the problem being solved here is the same as with the constrained column resize policy with non-integer scale JDK-8299753.  I was going to use some kind of simulated annealing algorithm there, perhaps it could be extracted into a generic utility and used in all three cases.

kevinrushforth commented:
This one hasn't been on my queue to review. I remember when I first took a look at it that it seems like a pretty large change that will need very careful review and a lot of testing. I seem to recall there are a couple places where the changes didn't seem quite right, but I'd need to go back and take a look.

@andy-goryachev-oracle raises a good point about perhaps wanting to solve this in a similar way for the various places where we need to do some sort of iterative adjustment (such as here and in the eventual solution for JDK-8299753). It seems also worth looking at what @Maran23 proposes for AnchorPane to fix JDK-8295078 in PR #910  (for which the review still needs to be completed).

Maran23 commented:
> @andy-goryachev-oracle raises a good point about perhaps wanting to solve this in a similar way for the various places where we need to do some sort of iterative adjustment (such as here and in the eventual solution for JDK-8299753). It seems also worth looking at what @Maran23 proposes for AnchorPane to fix JDK-8295078 in PR #910 (for which the review still needs to be completed).

Yes, at this point we should probably discuss what would be the best solution to fix the snapping issues once and for all.
We might even want to look into other frameworks like Swing. 
Maybe there are people internally at Oracle who can give valuable input here too?

hjohn reviewed (commented):
I only looked at HBox changes:

Although this code is probably an improvement, I'm not convinced it always is doing the right thing with regards to the final child sizes -- I think the container will indeed always fill up perfectly, but I think the child sizes can be off by a few pixels from ideal values given some crafted input values, that would require multiple passes in the adjust loop and multiple "snap" calls being applied.

Extending the tests to verify the final layout widths of the children could alleviate these doubts, as well as adding more edge cases.  I don't think the current tests exercise all of the newly written code; a coverage check could verify this.

I think the tests could be written to use an unsnapped simplistic sizing algorithm, and verifies resulting child widths are never off by more than 1 pixel.  It may also be possible to set the container to unsnapped, then to snapped and verify sizes never change by more than 1 pixel for random input values.

hjohn commented:
> > @andy-goryachev-oracle raises a good point about perhaps wanting to solve this in a similar way for the various places where we need to do some sort of iterative adjustment (such as here and in the eventual solution for JDK-8299753). It seems also worth looking at what @Maran23 proposes for AnchorPane to fix JDK-8295078 in PR #910 (for which the review still needs to be completed).
> 
> Yes, at this point we should probably discuss what would be the best solution to fix the snapping issues once and for all. We might even want to look into other frameworks like Swing. Maybe there are people internally at Oracle who can give valuable input here too?

I think extracting the common code out of HBox and VBox would be good, as they're basically doing the same thing in a different axis.  The code could then be unit tested directly (ie. given a set of min/max/pref sizes, a target size and a "snapper", calculate optimal sizes).  Something like this:

     double[] distribute(double space, double[] minWidths, double[] maxWidths, Snapper snapper);

The minWidths/maxWidths would be min + pref or pref + max depending on the situation, but this function wouldn't care.  The `Snapper` here would be aware of the scale, and would be a dummy implementation that does nothing when snapping is disabled.

This purposely does not handle spacing between children, as you can calculate that before hand and adjust `space` to the actual space available for children, simplifying the code.

Disclaimer: I used to create my own layout manager (see: [smartlayout](https://github.com/hjohn/hs.ui.smartlayout/blob/master/src/main/java/hs/smartlayout/LayoutRequirements.java)) that took into account min/max and weight of each child, and even groups of children (a restriction over two or more children at once) -- weight would affect resizing, allowing you to make layouts where children took space according to some ratio (1:2:3 for example).  This functionality I called a `SpaceDistributor` which was purposely given a neutral name as it could be used for both horizontal or vertical calculations.  When including weight, the calculations could get so complex that I had multiple implementations and an exhaustive search implementation (which was used by unit tests to verify the optimized implementations). This was before scaling became an issue though, so this code did not take snapping values to pixels into account.

andy-goryachev-oracle commented:
I suspect John is right: the kind of scenarios that would fail with the proposed code is when we have
a) fractional scale and
b) min/max constraints
c) requirement to fill the available space

The reason it would fail is that the component width is not invariant in respect of translation due to snapping and fractional scale.  Not only it means that multiple passes are required, but also that each time something gets changed or even moved, the computation needs to be discarded and re-done.  It is essentially the same issue as https://bugs.openjdk.org/browse/JDK-8299753

I think John is also right when talking about the common code.  I too, have encountered the problem with the Tree/TableView constrained resize policies (bug mentioned above) as well as with my own table layout implementation
https://github.com/andy-goryachev/FxDock/blob/master/doc/CPane.md

Perhaps we could come up with something along the lines John described

double[] distribute(double space, double[] prefWidth, double[] minWidths, double[] maxWidths, Snapper snapper);

though I would think the gaps between the components would have to be accounted for - again, due to snapping and fractional scale (or simply passed as components of the minWidths/maxWidths arrays.

I still owe you guys the formal review - mostly my intent is to hand-craft a case that would illustrate the problem.  It is still on my list of things to do.

cheers,
-andy




From: John Hendrikx ***@***.***>
Date: Wednesday, March 29, 2023 at 02:50
To: openjdk/jfx ***@***.***>
Cc: Andy Goryachev ***@***.***>, Mention ***@***.***>
Subject: [External] : Re: [openjdk/jfx] 8264591: HBox/VBox child widths pixel-snap to wrong value (#445)

@hjohn commented on this pull request.

I only looked at HBox changes:

Although this code is probably an improvement, I'm not convinced it always is doing the right thing with regards to the final child sizes -- I think the container will indeed always fill up perfectly, but I think the child sizes can be off by a few pixels from ideal values given some crafted input values, that would require multiple passes in the adjust loop and multiple "snap" calls being applied.

Extending the tests to verify the final layout widths of the children could alleviate these doubts, as well as adding more edge cases. I don't think the current tests exercise all of the newly written code; a coverage check could verify this.

I think the tests could be written to use an unsnapped simplistic sizing algorithm, and verifies resulting child widths are never off by more than 1 pixel. It may also be possible to set the container to unsnapped, then to snapped and verify sizes never change by more than 1 pixel for random input values.

________________________________

In modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java<https://urldefense.com/v3/__https:/github.com/openjdk/jfx/pull/445*discussion_r1151549496__;Iw!!ACWV5N9M2RV99hQ!KzBy-Nwzdqqe5UkuUWmRz2ReYMEkKWOyZ0g9cDGrsm88iZTdWLOyJ9It_9xcXBR2MwFW5YxmrL31nR_pgSDio97IfcEqEA$>:

>

-        double available = extraWidth; // will be negative in shrinking case

-        outer:while (Math.abs(available) > 1 && adjustingNumber > 0) {

-            final double portion = snapPortionX(available / adjustingNumber); // negative in shrinking case

-            for (int i = 0, size = managed.size(); i < size; i++) {

-                if (temp[i] == -1) {

+    /**

+     * Resizes the children widths to fit the target width, while taking into account the resize limits

+     * for each child (their minimum and maximum width). This method will be called once when shrinking,

Perhaps clarify why only once is sufficient for shrinking (if I read the code correctly, I think the reason is that adjustments always use the preferred children widths as a "base" value; the naming however is often generic enough that it could be their current layout widths that are being adjusted to a new reality, in which case even shrinking would need to take Priority into acount).

________________________________

In modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java<https://urldefense.com/v3/__https:/github.com/openjdk/jfx/pull/445*discussion_r1151582955__;Iw!!ACWV5N9M2RV99hQ!KzBy-Nwzdqqe5UkuUWmRz2ReYMEkKWOyZ0g9cDGrsm88iZTdWLOyJ9It_9xcXBR2MwFW5YxmrL31nR_pgSDio97De-abfA$>:

>      }



-    private double growOrShrinkAreaWidths(List<Node>managed, double areaWidths[][], Priority priority, double extraWidth, double height) {

-        final boolean shrinking = extraWidth < 0;

-        int adjustingNumber = 0;

+    /**

+     * Shrinks all children widths to fit the target width.

+     * In contrast to growing, shrinking does not require two phases of processing.

I find the terminology used confusing in many areas. Children widths seems to be referring to their preferred widths here, not (for example) their current widths. As these are preferred widths, it makes sense that shrinking would not need to look at priorities; if they had been current widths, you'd need to shrink SOMETIMES nodes first.

Documenting the parameter could alleviate this confusion. It may also be good to re-iterate in every parameter documentation whether this is a snapped value or not.

In fact, reading this a bit more, I think the "input" childrenWidths are pref + unused values (splitting the array would make that clearer). The values are snapped already. The empty values are then filled with minimum values, effectively modifying the array (should be documented), while the pref values are modified to become the layout values.

It looks like the double[][] construct is an optimization to prevent allocating two separate arrays... pretty sure that it would not make any difference if it had been split.

________________________________

In modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java<https://urldefense.com/v3/__https:/github.com/openjdk/jfx/pull/445*discussion_r1151589527__;Iw!!ACWV5N9M2RV99hQ!KzBy-Nwzdqqe5UkuUWmRz2ReYMEkKWOyZ0g9cDGrsm88iZTdWLOyJ9It_9xcXBR2MwFW5YxmrL31nR_pgSDio97K4wqIsg$>:

> @@ -467,77 +472,170 @@ private double[][] getAreaWidths(List<Node>managed, double height, boolean minim

         return temp;

     }



-    private double adjustAreaWidths(List<Node>managed, double areaWidths[][], double width, double height) {

+    /**

+     * Adjusts the children widths (within their min-max limits) to fit the provided space.

+     * This is necessary when the HBox is constrained to be larger or smaller than the combined preferred

+     * widths of its children. In this case, we grow or shrink the children until they fit the HBox exactly.

+     *

+     * @return the pixel-snapped content width, which is the combined width

+     *         of all children as well as the spacing between them

+     */

+    private double adjustChildrenWidths(List<Node> managed, double[][] childrenWidths, double width, double height) {

Could you document all the parameters, and mention what they represent (min, max, pref) and whether they're snapped or not?

Further, could you split childrenWidths in its constituent parts (I think it is prefWidths and maxWidths ?) -- the double[][] is confusing and after its initial creation, it can be passed as two separate arrays to subsequent functions for clarity.

________________________________

In modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java<https://urldefense.com/v3/__https:/github.com/openjdk/jfx/pull/445*discussion_r1151612290__;Iw!!ACWV5N9M2RV99hQ!KzBy-Nwzdqqe5UkuUWmRz2ReYMEkKWOyZ0g9cDGrsm88iZTdWLOyJ9It_9xcXBR2MwFW5YxmrL31nR_pgSDio94JmcG_nQ$>:

>

-        double[] usedWidths = areaWidths[0];

-        double[] temp = areaWidths[1];

-        final boolean shouldFillHeight = shouldFillHeight();

+        adjustWidthsWithinLimits(managed, usedWidths, minWidths, targetWidth, managed.size());

+    }

+

+    /**

+     * Grows all children widths to fit the target width.

+     * Growing is a two-phase process: first, only children with ***@***.*** Priority#ALWAYS} are eligible

+     * for adjustment. If the first adjustment didn't suffice to fit the target width, children with

+     * ***@***.*** Priority#SOMETIMES} are also eligible for adjustment.

+     */

+    private void growChildrenWidths(List<Node> managed, double[][] childrenWidths, double targetWidth, double height) {

+        double[] currentWidths = childrenWidths[0];

This is named usedWidths in the shrink version, while I think it actually are the preferred widths, which will be adjusted in this function to become the final layout widths.

As mutating values in Java is rare, a comment would help to clarify this:

double[] currentWidths = childrenWidths[0];  // initially pref widths, adjusted to layout widths

________________________________

In modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java<https://urldefense.com/v3/__https:/github.com/openjdk/jfx/pull/445*discussion_r1151650961__;Iw!!ACWV5N9M2RV99hQ!KzBy-Nwzdqqe5UkuUWmRz2ReYMEkKWOyZ0g9cDGrsm88iZTdWLOyJ9It_9xcXBR2MwFW5YxmrL31nR_pgSDio94S7vhfPQ$>:

> +    private boolean adjustWidthsWithinLimits(

+            List<Node> managed, double[] currentWidths, double[] limitWidths, double targetWidth, int adjustingNumber) {

+        double totalSpacing = (managed.size() - 1) * snapSpaceX(getSpacing());

+

+        // Current total width and current delta are two important numbers that we continuously

+        // update as this method converges towards a solution.

+        double currentTotalWidth = snapSpaceX(sum(currentWidths, managed.size())) + totalSpacing;

+        double currentDelta = targetWidth - currentTotalWidth;

+

+        // We repeatedly apply the following algorithm as long as we have space left to

+        // distribute (currentDelta), as well as children that are eligible to grow or

+        // shrink (adjustingNumber).

+        while ((currentDelta > Double.MIN_VALUE || currentDelta < -Double.MIN_VALUE) && adjustingNumber > 0) {

+            // The amount of space that, in the ideal case, we need to add to or subtract from

+            // each eligible child in order to fit the children into the target width.

+            double idealChange = snapPortionX(currentDelta / adjustingNumber);

Can this be 0 ?

For example, I have a targetWidth of 100, and two children that have preferred widths of 49 and 50, both with Priority.ALWAYS. I would need to grow to reach the target width of 100. As far as I can see the values in the code would then be:

 targetWidth = 100.0;

 currentTotalWidth = 99.0;

 currentDelta = 1.0;

 adjustingNumber = 2;

This would enter the while loop, and ideal change would be:

 idealChange = snapPortionX(1.0 / 2) = 0  // snapPortionX does a "floor".

That doesn't look good.

However, I now see that it wouldn't become an infinite loop, and that's because the input parameter adjustingNumber is being modified, and so doesn't always represent what I think it does.

________________________________

In modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java<https://urldefense.com/v3/__https:/github.com/openjdk/jfx/pull/445*discussion_r1151656708__;Iw!!ACWV5N9M2RV99hQ!KzBy-Nwzdqqe5UkuUWmRz2ReYMEkKWOyZ0g9cDGrsm88iZTdWLOyJ9It_9xcXBR2MwFW5YxmrL31nR_pgSDio97OrRRKqQ$>:

> +                currentWidths[i] = snapSizeX(oldWidth + actualChange);

+                currentTotalWidth = snapSpaceX(currentTotalWidth + (currentWidths[i] - oldWidth));

What I don't like about these two lines is that they're already dealing with values that are all snapped (and thus a fractional portion was lost). Then another calculation is done, and they're snapped again, losing even more information. Or is it that the snapping here is unnecessary as the calculation already involves values that are all have been snapped before?

So either, the snapping isn't needed, or things are being snapped multiple times -- the first is confusing, the second is like rounding values multiple times in a calculation instead of only the end result...

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/openjdk/jfx/pull/445*pullrequestreview-1362487101__;Iw!!ACWV5N9M2RV99hQ!KzBy-Nwzdqqe5UkuUWmRz2ReYMEkKWOyZ0g9cDGrsm88iZTdWLOyJ9It_9xcXBR2MwFW5YxmrL31nR_pgSDio965LQaXGA$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AZQ34ZECB4CKOP5B246PPW3W6QAVJANCNFSM42AIJ57Q__;!!ACWV5N9M2RV99hQ!KzBy-Nwzdqqe5UkuUWmRz2ReYMEkKWOyZ0g9cDGrsm88iZTdWLOyJ9It_9xcXBR2MwFW5YxmrL31nR_pgSDio97M0GlT9w$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

mstr2 commented:
Thank you very much @hjohn for your extensive review. Before addressing your comments, I'd like to get some clarification (ideally also from @kevinrushforth), because I am a bit confused by the discussion so far.

This PR is a very narrow patch that fixes an obvious bug in `HBox` and `VBox`. It's narrow in that it doesn't change the present algorithm, it merely corrects some of its flaws. There is a little bit of refactoring, but that's only done where it helps readers to better understand the algorithm.

However, the discussion seems to center around the idea of large-scale refactoring, even across multiple components, including  open tickets like [8299753](https://bugs.openjdk.org/browse/JDK-8299753). That's not something I can do as part of this PR, and if large-scale refactoring is the way we choose to go forward, I'd rather not spend more of my time on this particular PR. There needs to be a realistic chance that this PR will be accepted for integration basically as it is.

If we want to expand the scope of this work, this PR should be closed and the discussion should be continued on the mailing list.

andy-goryachev-oracle commented:
(copying here from the mailing list)

I suspect @hjohn  is right: the kind of scenarios that would fail with the proposed code is when we have
a) fractional scale and
b) min/max constraints
c) requirement to fill the available space
 
The reason it would fail is that the component width is not invariant in respect of translation due to snapping and fractional scale.  Not only it means that multiple passes are required, but also that each time something gets changed or even moved, the computation needs to be discarded and re-done.  It is essentially the same issue as https://bugs.openjdk.org/browse/JDK-8299753
 
I think John is also right when talking about the common code.  I too, have encountered the problem with the Tree/TableView constrained resize policies (bug mentioned above) as well as with my own table layout implementation
https://github.com/andy-goryachev/FxDock/blob/master/doc/CPane.md
 
Perhaps we could come up with something along the lines John described
 
`double[] distribute(double space, **double[] prefWidth**, double[] minWidths, double[] maxWidths, Snapper snapper);`
 
though I would think the gaps between the components would have to be accounted for - again, due to snapping and fractional scale (or simply passed as components of the minWidths/maxWidths arrays.

hjohn commented:
> Thank you very much @hjohn for your extensive review. Before addressing your comments, I'd like to get some clarification (ideally also from @kevinrushforth), because I am a bit confused by the discussion so far.
> 
> This PR is a very narrow patch that fixes an obvious bug in `HBox` and `VBox`. It's narrow in that it doesn't change the present algorithm, it merely corrects some of its flaws. There is a little bit of refactoring, but that's only done where it helps readers to better understand the algorithm.

Yes, I think it's an improvement in filling the boxes properly (I asked for some clarification as I found the code hard to read, but that's probably how you found it).  My only worry is that it fixes this specific problem, but may be introducing a new issue (I think snapping values multiple times can introduce artifacts in the child sizes, even though they will all align correctly, I think they might be off a bit).  If you can confirm this isn't the case, or if it is, fix those last problems then I think it is a positive change.

I have noticed similar problems on my system where I don't use 100 or 200% scaling, and it would be good to see them fixed.
 
> However, the discussion seems to center around the idea of large-scale refactoring, even across multiple components, including open tickets like [8299753](https://bugs.openjdk.org/browse/JDK-8299753). That's not something I can do as part of this PR, and if large-scale refactoring is the way we choose to go forward, I'd rather not spend more of my time on this particular PR. There needs to be a realistic chance that this PR will be accepted for integration basically as it is.

Those were possible future directions I suppose, not necessarily as part of this PR.  I dislike the code duplication, but it was that way already.

kevinrushforth commented:
> > Thank you very much @hjohn for your extensive review. Before addressing your comments, I'd like to get some clarification (ideally also from @kevinrushforth), because I am a bit confused by the discussion so far.

If we can resolve all of the concerns with the existing approach, without the need for a more holistic fix, then it seems OK to proceed with this. The question is: can we?

> > This PR is a very narrow patch that fixes an obvious bug in `HBox` and `VBox`. It's narrow in that it doesn't change the present algorithm, it merely corrects some of its flaws. There is a little bit of refactoring, but that's only done where it helps readers to better understand the algorithm.
> 
> Yes, I think it's an improvement in filling the boxes properly (I asked for some clarification as I found the code hard to read, but that's probably how you found it). My only worry is that it fixes this specific problem, but may be introducing a new issue (I think snapping values multiple times can introduce artifacts in the child sizes, even though they will all align correctly, I think they might be off a bit). If you can confirm this isn't the case, or if it is, fix those last problems then I think it is a positive change.

This was one my worries as well, but I haven't looked at it closely enough to point out any specific concerns.

andy-goryachev-oracle commented:
@mstr2 could you please merge in the latest master - the main line deviated a bit, it is difficult to test.  thank you!

andy-goryachev-oracle commented:
Any configuration where we have a fractional screen scale and setHGrow(true) is going to be a tricky one.  I don't think a single pass algorithm would ever work because the usual invariants (sum of all widths, snapped width) might change if even one component is moved or resized.

Here is an example where the algorithm fails - I am using the MonkeyTester app https://github.com/openjdk/jfx/pull/1097 - notice the last rectangle goes outside of the available width.

BTW, thanks for showing us how to visualize any possible errors in layout - I shamelessly stole your idea for the Monkey Tester.

![Screenshot 2023-04-21 161008](https://user-images.githubusercontent.com/107069028/233748074-5be88fb1-ee57-44ac-a753-53cedca1d76f.png)

hjohn commented:
I think a single pass algorithm should be possible. 

The fractional screen scale should make no difference here as you can treat sizes (after scaling when snapping is on) as integer values.  It is then a matter of assigning an integer number to each of the children.

What is clear to me however that the algorithm is sufficiently complex that it warrants direct unit testing. Verifying adjustments with manual testing is inevitably going to miss regressions as there are a lot of cases to take into account.

I've been testing out a separate function to distribute space with unit tests to verify its behavior, and it is quite similar to what already exists requiring only a single pass. The problem with the current algorithm is more that it is tightly integrated making it hard to test, not that it can't be made to function correctly.

hjohn commented:
I've put some additional time into this, and I can only conclude that when snapping is on, that every float calculation that is performed (even something as simple as an addition) must be rounded to pixel boundaries or you risk propagating a tiny error that will result in a comparison going the other way in some edge case.

I've now implemented an alternative in #1111 and to get all tests to perform exactly right and exactly when you'd expect to see a pixel added here or there, a lot of pixel boundary rounding is a requirement... so much so that I wonder if it really is wise to do this with floats -- it may be wiser to multiply everything by render scale first, do the calculations with integers, then restore it to floats.  I added @mstr2 's test to this one as well, and it passed unchanged.

andy-goryachev-oracle commented:
> I think a single pass algorithm should be possible.

In some simple cases, yes.  But not in all possible cases.

Let's consider one example, a layout similar to TableView.CONSTRAINED_RESIZE_POLICY_ALL_COLUMNS.  When we need to distribute extra space and at least one column (or a Node in a layout) hits its maximum size constraint, all the computations must be discarded and re-done using the maximum width for that node.  Since now we have more space to re-distribute, another node might hit its maximum constraint, and so the process must be repeated.

The situation is even worse when we have a fractional scale: the computation must be discarded and re-done not only if one node hits a constraint, but even when it gets moved.  Moving a node may not only require us to change the (already computed) width of that component, but also change the remaining delta, which basically invalidates the whole thing.

And finally, I doubt it is possible to use the trick mentioned in the preceding comment - by converting to and computing in the scaled coordinates, for a simple reason - the constraints are set in the original, unscaled and unsnapped coordinates.  When you try to convert a constraint to scaled and snapped coordinates, the constraint is not an invariant, that is it changes depending on where the node is positioned.

hjohn commented:
> > I think a single pass algorithm should be possible.
> 
> In some simple cases, yes. But not in all possible cases.
> 
> Let's consider one example, a layout similar to TableView.CONSTRAINED_RESIZE_POLICY_ALL_COLUMNS. When we need to distribute extra space and at least one column (or a Node in a layout) hits its maximum size constraint, all the computations must be discarded and re-done using the maximum width for that node. Since now we have more space to re-distribute, another node might hit its maximum constraint, and so the process must be repeated.

But that's exactly what HBox already does.  And you don't need to discard anything, you just continue but skip nodes that have already reached maximum and distribute the remaining space.

> The situation is even worse when we have a fractional scale: the computation must be discarded and re-done not only if one node hits a constraint, but even when it gets moved. Moving a node may not only require us to change the (already computed) width of that component, but also change the remaining delta, which basically invalidates the whole thing.

The start position of the Node should not influence the width or space available, see below.

> And finally, I doubt it is possible to use the trick mentioned in the preceding comment - by converting to and computing in the scaled coordinates, for a simple reason - the constraints are set in the original, unscaled and unsnapped coordinates. When you try to convert a constraint to scaled and snapped coordinates, the constraint is not an invariant, that is it changes depending on where the node is positioned.

I'm sorry, I don't quite see what you mean here, perhaps this is unique to the TableView layout. The constraints are rounded up to the nearest pixel boundary (even the maximum one).  They must be as it would not be possible to render both the left edge and right edge of a control on a pixel boundary if constraints weren't rounded.  The position is also always on a pixel boundary. The start position of a control has no bearing on the rounding of the constraint to pixels, as pixels are all equally wide.

andy-goryachev-oracle commented:
> I'm sorry, I don't quite see what you mean here, perhaps this is unique to the TableView layout.

It is not unique.  The scenario I am trying to describe is when set[VH]Grow is true, possibly for multiple nodes.  With the fractional scale and snapping, the snapped coordinates are not equidistant.  So if one node has to move, all of a sudden its width must change, affecting the others.  Which means

1. one cannot scale the constraints as they become position-dependent
2. no single pass algorithm is going to work

(the reason  I mentioned TableView is because I've tripped over the very same problem, see JDK-8299753)

For reference:
```
scale=1.25
x=1 snapped=0.800 step=0.800
x=2 snapped=2.400 step=1.600
x=3 snapped=3.200 step=0.800
x=4 snapped=4.000 step=0.800
x=5 snapped=4.800 step=0.800
x=6 snapped=6.400 step=1.600
x=7 snapped=7.200 step=0.800

scale=1.5
x=1 snapped=1.333 step=1.333
x=2 snapped=2.000 step=0.667
x=3 snapped=3.333 step=1.333
x=4 snapped=4.000 step=0.667
...
```

hjohn commented:
> Perhaps i did not emphasized one thing: I believe single pass algorithms would not work when all of the following are true:
> 
> * min/max constraint(s)
> * grow property(ies) are set
> * fractional scale
> * snapping
> 
> But then again, perhaps I do miss some fundamental aspect of the process, and it is possible. I welcome a PR than proves me wrong!

You can check out #1111 -- I'm not aware of any problems in it.

hjohn commented:
> > I'm sorry, I don't quite see what you mean here, perhaps this is unique to the TableView layout.
> 
> It is not unique. The scenario I am trying to describe is when set[VH]Grow is true, possibly for multiple nodes. With the fractional scale and snapping, the snapped coordinates are not equidistant. So if one node has to move, all of a sudden its width must change, affecting the others. Which means
> 
> 1. one cannot scale the constraints as they become position-dependent
> 2. no single pass algorithm is going to work
> 
> (the reason I mentioned TableView is because I've tripped over the very same problem, see JDK-8299753)
> 
> For reference:
>
> scale=1.5
> x=1 snapped=1.333 step=1.333
> x=2 snapped=2.000 step=0.667
> x=3 snapped=3.333 step=1.333
> x=4 snapped=4.000 step=0.667

Okay, I'm going to explain my understanding of the system, please correct me where I'm going wrong:

We have logical positions and sizes, which are `double`s and can be any value, not just integers. And we have physical positions and sizes which can also be any value, but can be snapped if we want them to correspond to physical pixels.  These snapped physical coordinates will be integers, and when converted back to logical ones will become fractional for most render scales.

To go from logical to physical, you multiply the logical value by render scale.  To go from physical to logical, you divide the physical value by render scale.  To "snap" a logical value, you multiple it by the render scale, round it (depending on type), then divide it by render scale. To "snap" a physical value, you round it to an integer.

|Scale|Logical|Physical|Logical (snapped)|Phyiscal (snapped)|
|---|---|---|---|---|
|1.0|0.0|0.0|0.0|0.0|
|1.0|0.5|0.5|1.0|1.0|
|1.0|1.0|1.0|1.0|1.0|
|1.0|1.5|1.5|2.0|2.0|
|1.5|0.0|0.0|0.0|0.0|
|1.5|0.5|0.75|0.66|1.0|
|1.5|1.0|1.5|1.33|2.0|
|1.5|1.5|2.25|1.33|2.0|

`HBox` will determine positions itself, and is not limited to integer logical coordinates, see example:

Given a scale of 1.5, and HBox has a logical margin of 1, and snapping is on.  After snapping, the margin becomes a logical width of 1.33 (2 pixels wide) [calculation: round(1.0 * renderScale) / renderScale].  The first control therefore gets positioned at a logical position of 1.33.  It's logical width can take on any multiple of the physical pixel size divided by the render scale, which is 1 / 1.5 = 0.66.  So it can span from 1.33 to 2.0 or 1.33 to 2.66 or 1.33 to 3.33, etc.

What am I missing?

Are the column widths in TableView perhaps restricted to integer values?

andy-goryachev-oracle commented:
> It's logical width can take on any multiple of the physical pixel size divided by the render scale, which is 1 / 1.5 = 0.66.

No it cannot.  Snapping imposes limitation on which logical coordinates are acceptable.  What's more, the distance between two adjacent acceptable logical coordinates is not the same.

Let's take a look at an example.  We are going to use logical coordinates only, because these are used for constraints as well as in the layout.  Here is a list of acceptable snapped logical coordinates for scale=1.5 again:

```
scale=1.5
x=1 snapped=1.333 step=1.333
x=2 snapped=2.000 step=0.667
x=3 snapped=3.333 step=1.333
x=4 snapped=4.000 step=0.667
x=5 snapped=5.333 step=1.333
```

(the fact that x are integers is irrelevant, we are not limited to integers, it's just to illustrate a point).

So let's have two nodes with fixed size of 1 (min = max = 1).  The first one is positioned at x=0 and has the width of 1.333.
The second is positioned at x=1.333 and has a width of 0.667.

One surprising effect of this (that we haven't discussed yet) is that both min and max constraints cannot be honored (this might impact some tests).

But let's move to the main point.  Imagine that we have more nodes, and as the layout progresses, we discover that one or more nodes hit the max constraint.  And let's imagine that the node that gets extra space is the first node.  So the first node gets wider (width = 2.0), then the second node is moved and its width must change to 1.333.  This means the width (already computed earlier) in not an invariant, it may change, and in doing so pretty much everything else needs to be recomputed.  And if in turn any nodes hit (or stop hitting) their constraints, we may need to do it again and again.  Could we have an endless loop condition where the layout goes between two (or any other number of) possible states?

To summarize, snapping and fractional scale under certain circumstances (*) leads to:

- honoring constraints might be impossible
- moving a single node requires a new layout pass
- possible danger of infinite loop cycling between 2+ layout states

(*) the circumstances are: nodes having grow property set.

hjohn commented:
> > It's logical width can take on any multiple of the physical pixel size divided by the render scale, which is 1 / 1.5 = 0.66.
> 
> No it cannot. Snapping imposes limitation on which logical coordinates are acceptable. What's more, the distance between two adjacent acceptable logical coordinates is not the same.

Snapping turns physical coordinates into integers as you can't have half pixels.  Pixels are uniformly spaced.  Converting a pixel offset back to a logical coordinate leads to a fractional value, but any pixel offset can be converted back, and the snapped logical coordinates will be uniformly spaced because of that.
 
> Let's take a look at an example. We are going to use logical coordinates only, because these are used for constraints as well as in the layout. Here is a list of acceptable snapped logical coordinates for scale=1.5 again:
> 
> ```
> scale=1.5
> x=1 snapped=1.333 step=1.333
> x=2 snapped=2.000 step=0.667
> x=3 snapped=3.333 step=1.333
> x=4 snapped=4.000 step=0.667
> x=5 snapped=5.333 step=1.333
> ```

What about:
```
x = 0.5 snapped = 0.666
x = 2.5 snapped = 2.666
```
?

> (the fact that x are integers is irrelevant, we are not limited to integers, it's just to illustrate a point).
> 
> So let's have two nodes with fixed size of 1 (min = max = 1). The first one is positioned at x=0 and has the width of 1.333. The second is positioned at x=1.333 and has a width of 0.667.

Why would the 2nd node have a width of 0.667?  That would be below its minimum allowed value.  It will be 1.333 just like the first node.  The container will ensure that its minimum value takes the snapping into account, so it will report a logical minimum of 2.666 (ie. exactly 4 pixels).

The min/max values are not strictly honored when snapping is on, they're snapped first.  So `min = max = 1` would become `min = max = ceil(1 * 1.5) / 1.5 = 1.333`.  It's therefore possible that the minimum size of a control won't reach the logical minimum value if that minimum is not an exact number of pixels.  The maximum can even be exceeded by up to 1/1.5.

> To summarize, snapping and fractional scale under certain circumstances (*) leads to:
> 
> * honoring constraints might be impossible
> * moving a single node requires a new layout pass
> * possible danger of infinite loop cycling between 2+ layout states
> 
> (*) the circumstances are: nodes having grow property set.

I only agree with the first point, because constraints are snapped and therefore may be slightly different than their logical values, but this rule applies in all cases:

     snapped constraint >= logical constraint

... and, as constraints are spaces, they are always rounded up when snapping.

hjohn commented:
@mstr2 I noticed in your test that you said that when `prefWidth` is set, the container must use exactly that size, but this is not entirely true. It just so happens that `76.0` is a pixel boundary for all the render scales you tested.  If you use a different width or use a different render scale (like `1.33333333`) then it won't be a pixel boundary and the width will not be exactly what you specified for preferred width.  With pref width `76.0` and render scale `1.33333333` the final width will be `76.5` for example.

andy-goryachev-oracle commented:
> The min/max values are not strictly honored when snapping is on, they're snapped first.

[question]

Here is a general question: in the case when min==max, in which direction should we snap?  Could this question be answered all the time, or do we need some sort of hint from the app developer?

For example, we may decide to honor min width over max width when snapping, to avoid text overruns in labels and buttons.  At the same time, there might be situations where max width is more important, I just can't think of any.

hjohn commented:
> > The min/max values are not strictly honored when snapping is on, they're snapped first.
> 
> [question]
> 
> Here is a general question: in the case when min==max, in which direction should we snap? Could this question be answered all the time, or do we need some sort of hint from the app developer?
>
> For example, we may decide to honor min width over max width when snapping, to avoid text overruns in labels and buttons. At the same time, there might be situations where max width is more important, I just can't think of any.

In UI layout, it's pretty standard that min always wins over max, and that's also what JavaFX is doing (see the `boundedSize` function in `Region` that handles this).  Here is how it is called when calculating the max value for `HBox`:

        // if min > max, min wins, so still need to call boundedSize()
        return left + snapSizeX(boundedSize(child.minWidth(alt), max, Double.MAX_VALUE)) + right;

The above basically returns `max` only if it is within the given range of `minWidth` and `Double.MAX_VALUE`, otherwise it is clamped to one of those two values.

Also, the developer already has full control over the behavior. If the snapping is not to their liking, it can be disabled and values can be manually calculated using the public snapping functions or by querying render scale themselves. I see no need to complicate this already complicated process even further as the current behavior is I think what the most developers will expect (ie. content sizes are ceil'd so you always get at least 1 pixel (unless 0), and spaces are rounded as they're less important).

andy-goryachev-oracle commented:
[question]

After giving it some thought, I agree with @hjohn that a single pass algorithm is possible if we do *everything* in snapped coordinates, and accept the fact that min/max constraints might get violated.

Do we want to continue with this PR, or switch to https://github.com/openjdk/jfx/pull/1111 ?

kevinrushforth commented:
PR  #1111 is likely to be the eventual solution for the general problem of handling layout when snap-to-pixel is set (as it is by default). So the question I have, primarily for @mstr2 and @hjohn, is: Is this PR a reasonable step along the path to where we want to go? Or will it need to be reworked or reverted if and when we get to PR #1111 ?

If it's the former, it might make sense to revive this PR and take it forward. Otherwise, it should probably be closed in favor of continuing work on PR #1111

hjohn commented:
> PR #1111 is likely to be the eventual solution for the general problem of handling layout when snap-to-pixel is set (as it is by default). So the question I have, primarily for @mstr2 and @hjohn, is: Is this PR a reasonable step along the path to where we want to go? Or will it need to be reworked or reverted if and when we get to PR #1111 ?
> 
> If it's the former, it might make sense to revive this PR and take it forward. Otherwise, it should probably be closed in favor of continuing work on PR #1111

It could perhaps be a reasonable step, although most of the effort would become wasted if we also move #1111 forward.  I'd expect a good JUnit test though to prove that it works correctly.  It's also possible that the integration of #1118 already solved some of the problems (but I'm sure it did not solve all of them), and so this PR and #1111 may have become less important.

mstr2 commented:
I've tested #1111, which solves the problem that this PR addresses.
I've also tested #1118, which is not sufficient to fix the problem (child widths can still exceed container width).
Since the implementation is quite different, I think we should close this PR, and move #1111 forward.

> It's also possible that the integration of https://github.com/openjdk/jfx/pull/1118 already solved some of the problems (but I'm sure it did not solve all of them), and so this PR and https://github.com/openjdk/jfx/pull/1111 may have become less important.

The main problem is still present with #1118, so a solution for that has not become less important.
